### PR TITLE
discv5

### LINKED ISSUE #15: Research: Discv5 feasibility
The purpose of this feasibility study is to investigate `discv5`s ability to function efficiently on resource restricted devices. The goal is to be able to judge both the bandwidth consumption and the ability to function within small windows of connectivity.

## Acceptance Criteria

This feasibility study looks at various problems:
 - bandwidth consumption
 - ability to function in small windows of connectivity
 - needle in haystack (how hard is it to find a node with certain characteristics)

It is supposed to provide insight on all of these problems allowing for us to form a conclusion on `discv5`s viability. 

## Method

The testing should be kept fairely simple, the configurable paramters are the following:

 - `nodes` - indicates the amount of nodes.
 - `needles` - indicates the amount of nodes which are `needles` in our `haystack` or set of nodes.
 - `window` - indicates the window size in seconds for the searching node to be online.

A discv5 DHT should, `n` amount of `nodes` should be generated and added to the DHT of which there are an amount of `nodes` that fit the `needle` description. A mobile client is simulated that connects to the DHT in short bursts indicated by `window`. Bandwidth should be recorded while connecting and disconnecting and synchronizing packets. It should also be checked how many round trips are required to find the specified `needles`.

## Notes

 - The `discv5` spec is currently in motion and always changing. Even though we have the [nim](https://github.com/status-im/nim-eth/tree/master/eth/p2p/discoveryv5) implementation it may make sense to use the [go-ethereum](https://github.com/ethereum/go-ethereum/tree/master/p2p/discv5) version for this reason.

### PR DESCRIPTION
**Pull request by decanus**
closes #15

### CODE REVIEW

**Comment by decanus** [comment]
So @oskarth, I was working a bit with @kdeme and we both had the same thought. I actually started googling and on most devices there is no practical way to get your external IP. This to me means that mobile devices cannot sign an ENR and participate in the DHT in 2 ways. They can only run lookups. So the only real experiment that has any form of practical effect is figuring out how many lookups a "light" node needs to execute to find interesting nodes, this is totally dependent on the ordering of the DHT and if we were to take into account what @arnetheduck said yesterday may not even be relevant. IMO topics is what is needed here, has the lowest overhead and always guarantees that we ideally get what we are looking for when doing 1 lookup.

**Comment by decanus** [comment]
So also talking to @kdeme, a node can reply with the external IP in a `PONG` message so people can update their ENRs. But considering the short connection windows + the fact that mobile gets new IP address much more often it is questionable if this will work well or even work at all.

**Comment by kdeme** [comment]
Yes, the  important part is to know how expensive (how many lookups) it will be to get n interesting nodes in a bunch of nodes (randomly filled DHT).
This will also depend on some of the parameters I mention here: https://github.com/vacp2p/research/issues/15#issuecomment-601657832
Obviously, topic advertisement is an important one there, and depending on the ratio of interesting versus general nodes, one could probably already conclude that it is necessary.

The fact of short connection windows + frequently updated ip-address will probably cause a mobile device not to be discoverable from other nodes in the DHT (chance of others finding a correct ENR in the DHT). At least, not when the mobile device is moving ( guess this also depends somewhat per mobile operator, ipv6, etc.).

**Comment by oskarth** [comment]
I'm not following. As far as I'm concerned it has been known for a long time that mobile phones wouldn't be publicly connectable. The purpose of the DHT isn't to find mobile phone, it is to find nodes that provide a useful service. The questions that should be answered are still the same.

**Comment by decanus** [comment]
@oskarth connection info in the ENR should be signed https://github.com/ethereum/devp2p/blob/master/enr.md. This includes IP etc, even though these fields are optional an ENR without them would be usefless.

**Comment by decanus** [comment]
So we seem to have a problem @kdeme, there are times when nodes just return nothing in the lookup.

File: discv5/src/discv5.nim [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+proc randNode(nodes: seq[discv5_protocol.Protocol]): Node =
+    randomize()
+    result = nodes[rand(N - 1)].localNode
+
+proc randNodeFromNodes(nodes: seq[Node]): Node =
+    randomize()
+    result = nodes[rand(nodes.len - 1)]
+
+proc runWith(node: discv5_protocol.Protocol, nodes: seq[discv5_protocol.Protocol]) {.async.} =
```
Comment (dryajov) [inline discv5/src/discv5.nim]: Could you please elaborate on the purpose of this method?

File: discv5/src/utils.nim [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    result = readUintBE[256](keccak256.digest(pk.getRaw()).data)
+
+proc containsNodeId*(s: seq[string], id: string): bool =
+    for n in s:
+        if n == id:
+            return true
+
+    return false
+
+proc distanceTo*(a, b: NodeId): uint32 =
```
Comment (oskarth) [inline discv5/src/utils.nim]: Is this the distance function used in Disc v5? If so, what's the reason for not using it directly?

File: discv5/src/discv5.nim [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            peer = randNodeFromNodes(lookup)
+
+    echo "Not found in max iterations"
+
+proc run() {.async.} =
+    var nodes = newSeq[discv5_protocol.Protocol](0)
+
+    for i in 0..<N:
+        let node = initDiscoveryNode(newPrivateKey(), localAddress(20300 + i), if i > 0: @[nodes[0].localNode.record] else: @[])
+        node.start()
```
Comment (oskarth) [inline discv5/src/discv5.nim]: How do these N nodes discover each other initially?

File: discv5/src/utils.nim (hunk 2 above)
Comment (kdeme) [reply discv5/src/utils.nim]: yeah, that looks like the `logDist` cakk from discv5, not sure why it was re-added here.

File: discv5/src/discv5.nim (hunk 3 above)
Comment (kdeme) [reply discv5/src/discv5.nim]: The first node acts as bootstrap for the N-1 others.
The `node.start()` line will start the lookup loop and revalidation loop. The first should run a self lookup and a random lookup every minute. However, I'm not so sure that with the 50 second sleep of below, every node of the N nodes actually manages to complete that. And even if it does, 1 run is perhaps a little low. So the DHT of some nodes might not be populated enough yet.

Comment (decanus) [reply discv5/src/discv5.nim]: @kdeme I also tried with 5 minutes, it didn't make much difference to the result.

Comment (oskarth) [reply discv5/src/discv5.nim]: Are you logging those nodes?

**Comment by kdeme** [comment]
So I ran the code and did a little investigation

First I noticed a lot of revalidation failures, and many of them are with the bootstrap node (makes sense, all are spamming this one initially). I also noticed a lot messages arriving after the timeouts (and thus being dropped).

I added some logging to see how many entries each node has in its DHT, it turned out to be very low for the nodes started later, see first file here:
https://gist.github.com/kdeme/e498f4a9efc1dbb21931f74648d2bcec

So I increased the timeouts (`handshakeTimeout = 5.seconds, responseTimeout = 10.seconds`). 
Results for the entries in the DHT seem better (see second file in gist), however, the lookups are still rather bad. And there are some strange cases where it needs several lookups or sometimes even fails when a node is located in 99 of the DHTs.

So, it looks like there is some bug in the findNode call, I will further investigate this.

Even with the increased timeouts, there are still a lot of messages that get lost. We could probably improve this by doing the lookups (that now happen in the loops) in a more controlled way. Or by adding them "manually" (yet randomly) with `addNode`, as we have all the nodes their information (ENR).

**Comment by decanus** [comment]
@kdeme @oskarth just posting this here for persistence, when I now start a network by pairing nodes randomly using the `addNode` function, ensuring that `start` function is called on no node other than the bootstrap node we seem to get a success rate more close to 60-80%, rather than a failure rate to that. This would mean that there is something wrong with the lookup loop used by the discv5 implementation in nim.

**Comment by oskarth** [comment]
Questions on baseline test results: https://gist.github.com/decanus/dea759e3471b0a5dcebba1218eb070c7

> PEERS_PER_NODE = 16
This is the setting, but is this the reality? That's what I would assert in log statements

> If a node returns 0 nodes when called with a distance of 256, we consider it to be failed. We limited max lookups to 100, it seems unreasonable to assume it would go higher, and none of the results seemed to go higher.

IIRC you saw behavior where a node return 0 nodes several times in a raw, but then it returned more. Is that issue fixed and do we know the origin of that?

> by pairing nodes randomly using the addNode function

The important property to me is that each node should have kademlia connectivity. This may or may not happen by adding random nodes. I'd make sure the nodes in the network have connectivity by logging/asserting that they should have a node in each k-bucket.

**Comment by kdeme** [comment]
So I've investigated the code for FindNode in nim-eth better and found a crucial bug.
Fix is here https://github.com/status-im/nim-eth/pull/219
More background here: https://github.com/status-im/nim-eth/issues/205#issuecomment-612095492
As you can read there, if I now also change the returned values to only nodes with the correct distance. Then I think we will see more lookups required.

I've ran (an older version, sorry didn't update) of your code and results are much better.
Failure can still occur of course but at least it seems more consistent (less nodes in the tables of the individual nodes -> more chance on failure, etc.)

**Pull request merged**